File: Tarea7/Codigo/calcVectFrec.py
"""
    Por I. Alejandro Gómez Pérez

    Esta es una adaptación del código con el mismo nombre visto en clase en lenguaje matlab a python

    Básicamente construye un arreglo con las frecuencias en el espacio de fourier ordenadas de 
    menor a mayor.

    #### Parámetros:
        - datos: array de la señal original
        - T_muestreo: Periodo de muestreo de los datos
    #### Retorna:
        - vfreq: un array con los elementos ordenados comenzando con los valores negativos pasando por 0 
        hasta los positivos para poder interpretar mejor la fft
"""

from numpy import arange,floor
from numpy.fft import ifftshift

def calcVecFrec(datos,PeriodoMuestreo):
    #Esta funcion calcula el vector de los valores de la frecuencia (en Hz) en el
    #espacio de Fourier de una secuencia guardada en el vector datos,
    #PeriodoMuest, es el tiempo transcurrido entre medición y medición
    N = len(datos) # Calculamos la longitud de los datos
    TiempoCompleto=PeriodoMuestreo*N #Calculamos el tiempo total (de medición) de la señal
    #Verificamos si la cantidad de datos es par o impar
    if(N%2 == 0):
        # Construímos un vector que va de -int(N/2) hasta int(N/2)-1 en incrementos de 1 en 1
        vfreq = arange(-floor(N/2),floor(N/2),1)
    else:
        # Construimos un vector que va de -int(N/2) hasta int(N/2) en incrementos de 1
        vfreq = arange(-floor(N/2),floor(N/2)+1,1)
    # Hasta este punto las unidades de vfreq serán (repeticiones)
    # Queremos convertirlas a Hz como en el caso de la fft, así que dividiremos
    # entre el tiempo completo de medicion
    vfreq /= TiempoCompleto # Aqui las unidades son Hertz
    # Corremos el arreglo de tal forma que quede identico a la parte del dominio de la gráfica de
    # la transformada de fourier

    # Lo que hace ifffshift es correr un arreglo hacia la izquierda por ejemplo:
    # si vfreq = [-2,-1,0,1,2] ---> [-1,0,1,2,-2] ---> [0,1,2,-2,-1]
    vfreq = ifftshift(vfreq) # Corremos el arreglo
    return vfreq

File: CodigosClase/Codigos_Matlab/CalcVectFrec.m
function [ vfrec ] = CalcVectFrec( datos, PeriodoMuest )

%Esta funcion calcula el vector de los valores de la frecuencia (en Hz) en el
%espacio de Fourier de una secuencia guardada en el vector datos,
%PeriodoMuest, es el tiempo transcurrido entre medici?n y medici?n
N=size(datos,2);
%se calcula el valor en hertz de la frecuencia m?s baja de una onda que estar?a 
%completa en el vector de datos que se analiza, vean que el TiempoCompleto se
%calcula como el numero de muestras por el tiempo entre muestras
TiempoCompleto=PeriodoMuest*N;
if mod(N,2)==0
    %si las mediciones son pares se aparta un lugar entre los positivos
    %para el 0, la frec absoluta m?s alta se encontrar? en la secci?n de
    %las frec negativas y sera el n?mero de muestras N dividido entre 2,
    %esto se debe al teorema del muestreo que dice que la frecuancia m?s
    %alta que se puede observar es la mitad de la frecuencia de muestreo.
    vfrec=-floor(N/2):1:floor(N/2)-1;
else
    %si el numero de muestras es impar, entionces hay lugar para el 0
    vfrec=-floor(N/2):1:floor(N/2);
end
%Hasta la l?nea anterior las unidades de la frecuencia son repeticiones por 
%el tiempo completo de medici?n, para convertirlas en hertz (repeticiones 
%por segundo) hay que dividir entre el tiempo completo 
vfrec=ifftshift(vfrec)/TiempoCompleto;
%ahora si las unidades de frecuencia est?n en hertz, recuerde que el
%significado del signo menos en la frecuencia es un simplce cambio de fase.
end

